File: LC1372-LongestZigzagPathInABinaryTree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int maxLen = 0;

// direction: 0 - left, 1 - right
void dfs(struct TreeNode* node, int direction, int length) {
    if (!node) return;

    if (length > maxLen) maxLen = length;

    // If we go the same direction, reset length to 1,
    // else increment length.
    dfs(node->left, 0, direction == 0 ? 1 : length + 1);
    dfs(node->right, 1, direction == 1 ? 1 : length + 1);
}

int longestZigZag(struct TreeNode* root) {
    maxLen = 0;
    // Start from root, go left or right
    dfs(root->left, 0, 1);
    dfs(root->right, 1, 1);
    return maxLen;
}


// DFS solution
// Time: O(n)
// Space: O(h) h is the height of the tree

File: LC1372-LongestZigzagPathInABinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.max_length = 0

        # direction: 0 for left, 1 for right
        def dfs(node, direction, length):
            if not node:
                return

            self.max_length = max(self.max_length, length)

            # If we go the same direction, reset length to 1,
            # else increment length.
            if direction == 0:
                dfs(node.left, 1, length + 1)
                dfs(node.right, 0, 1)
            else:
                dfs(node.right, 0, length + 1)
                dfs(node.left, 1, 1)

        dfs(root, 0, 0)
        dfs(root, 1, 0)

        return self.max_length